### PR TITLE
Consolidate sync/async JsonSerializer test abstractions

### LINKED ISSUE #66463: Consolidate sync/async JsonSerializer test abstractions into single type and expand async coverage for reference handler feature
Consolidate sync/async JsonSerializer test abstractions into single type and expand async coverage for reference handler feature.

From @eiriktsarpalis in https://github.com/dotnet/runtime/pull/66422#discussion_r823571429:

> I think all these tests should be made to target both stream-based and string-based serialization, since reference preservation logic contains bespoke async code.

>
> If anything, that's more feedback on the design of the base [SerializerTests](https://github.com/dotnet/runtime/blob/9030b78f19b363da4ffcaa00d8ae4929b010fd48/src/libraries/System.Text.Json/tests/Common/SerializerTests.cs#L8-L10) class. I don't think we should have separate entrypoints for "stream" and "string", since both can be abstracted behind a single serializer interface (effectively the JsonSerializerWrapperForString API). The different modalities can then be tested separately using ReferenceHandlerTests_Sync or ReferenceHandlerTests_Async derived classes that specify only one SUT.

>
> [We already do this in a few other places](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/JsonSerializerWrapperForString.cs#L123) but we're far from being consistent.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Our existing testing abstractions distinguish between streaming and non-streaming APIs (`JsonSerializerWrapperForString` nad `JsonSerializerWrapperForStream`), meaning that test authors need to explicitly write separate tests in order to exercise async serialization test coverage for their feature. This PR consolidates our testing infrastructure so that a single abstraction is used for both streaming and non-streaming APIs, getting async test coverage for free from our existing non-async unit tests.

More specifically this change:

1. Makes `JsonSerializerWrapperForStream` inherit from `JsonSerializerWrapperForString`, implementing string-based shims for the streaming APIs.
2. Makes the `SerializerTests` base class accept a single `JsonSerializerWrapperForString` parameter, defining a unique System Under Test for each test class.
3. Extends the `AsyncStreamSerializerWithSmallBuffer` serializer so that `DefaultBufferSize = 1` is enforced on custom `JsonSerializerOptions` arguments as well. This should let us stress test our async converters more comprehensively.
4. Misc renamings and quality of life improvements.

I've broken up the PR into two separate commits that should best be reviewed separately:

* https://github.com/dotnet/runtime/pull/66729/commits/31ae0f4bfb8dad6909352c438019f0d002fa8c4f contains the core infrastructural refactoring.
* https://github.com/dotnet/runtime/pull/66729/commits/4cdeca3ce0092ea6efcc37aa4f8ef415eede517e contains a couple of rename refactorings that impact the vast majority of test classes. Straightforward but verbose.

The following issues have been filed following this refactoring: #66727, #66687.

Fix #66463.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs [hunk 1]
```diff
@@ -18,18 +18,23 @@ public abstract partial class CollectionTests
...
                 DefaultBufferSize = bufferSize
             };
 
-            string expectedJson = await JsonSerializerWrapperForString.SerializeWrapper(source);
+            string expectedJson = JsonSerializer.Serialize(source);
 
             using var stream = new Utf8MemoryStream();
             var asyncEnumerable = new MockedAsyncEnumerable<TElement>(source, delayInterval);
-            await JsonSerializerWrapperForStream.SerializeWrapper(stream, asyncEnumerable, options);
+            await StreamingSerializer.Serialize(stream, asyncEnumerable, options);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs:35]: Past discussion around naming was that we didn't want to use the same method name "Serialize" and "Deserialize" as the `JsonSerializer` to make code searches easier. I don't have a strong opinion either way, but that's the reasoning behind the current names.

**Comment by steveharter** [comment]
> meaning that test authors need to explicitly write separate tests in order to exercise async serialization test coverage for their feature.

The pattern for tests that want to verify more than one main serializer API was to implement the tests as instance methods in an abstract class, then derive from that abstract class for the various permutations (stream, async, sync,...). So the tests did not have to be separate or duplicated. Does this PR change that and\or does it suggest a different pattern?

**Comment by eiriktsarpalis** [comment]
> > meaning that test authors need to explicitly write separate tests in order to exercise async serialization test coverage for their feature.
> 
> The pattern for tests that want to verify more than one main serializer API was to implement the tests as instance methods in an abstract class, then derive from that abstract class for the various permutations (stream, async, sync,...). So the tests did not have to be separate or duplicated. Does this PR change that and\or does it suggest a different pattern?

The change is doubling down on that pattern, correcting a few divergences that were likely accidental. In particular the `SerializerTests` base class was explicitly distinguishing between streaming and non-streaming tests which was effectively forcing segregated test coverage. Each abstract test class now only requires a single system under test, and tests needing to exercise streaming APIs directly [can rely on the runtime type of the SUT](https://github.com/dotnet/runtime/pull/66729/files#diff-734748330c3c1ede0bec766df2520037dd07ffff15b1696c3ff1a18f7279f62bR24).

File: src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs [hunk 2]
```diff
@@ -18,18 +18,23 @@ public abstract partial class CollectionTests
...
+                return;
+            }
+
             JsonSerializerOptions options = new JsonSerializerOptions
             {
                 DefaultBufferSize = bufferSize
             };
 
-            string expectedJson = await JsonSerializerWrapperForString.SerializeWrapper(source);
+            string expectedJson = JsonSerializer.Serialize(source);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs:31]: Why were these changed from wrappers to direct calls?

Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs:31]: I assume no desire to wrap since these are just for obtaining expected JSON for the actual test,

File: src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.IgnoreCycles.cs [hunk 3]
```diff
@@ -456,21 +456,11 @@ private async Task Test_Serialize_And_SerializeAsync<T>(T obj, string expected,
 
             if (objType != typeof(object))
             {
-                json = await JsonSerializerWrapperForString.SerializeWrapper(obj, options);
-                Assert.Equal(expected, json);
-
-                using var ms1 = new MemoryStream();
-                await JsonSerializerWrapperForStream.SerializeWrapper(ms1, obj, options).ConfigureAwait(false);
-                json = Encoding.UTF8.GetString(ms1.ToArray());
+                json = await Serializer.SerializeWrapper(obj, options);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.IgnoreCycles.cs:459]: So these can be combined now since this test class will run under both string+stream?

**Review by steveharter — APPROVED** [review]
A couple questions, but otherwise looks like a nice refactoring.

File: src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.AsyncEnumerable.cs:31]: Yep, that's the underlying motivation here.
